[WIP] Review existed api

f1xgun opened the pull request.

f1xgun committed:
fix(therapists): return 200 with empty list instead of 404

Changed GET /therapists to return empty array when no records exist
instead of throwing 404 error for better API consistency.

f1xgun committed:
fix(roles): return 200 with empty list instead of 404

Changed GET /roles to return empty array when no records exist
instead of throwing 404 error for better API consistency.

f1xgun committed:
feat(makefile): add database migration commands

Added convenient make targets for managing Alembic migrations:
- migrate: apply all pending migrations
- migrate-create: create new migration with autogenerate
- migrate-rollback: rollback last migration
- migrate-status: show current migration version
- migrate-history: show migration history

Both Docker and local execution modes supported.
Also fixed docker-compose command format (docker compose → docker-compose).

f1xgun committed:
feat(appointments): add scheduled_time, comment fields and refactor error handling

- Add scheduled_time (required) and comment (optional) to appointments
- Refactor services/appointments.py into modular package structure
- Replace generic ValueError with custom exceptions
- Add proper HTTP status codes (404 for not found, 422 for validation)
- Add timezone-aware datetime validation
- Use match/case for appointment type handling

Breaking change: API now requires scheduled_time field for appointments

f1xgun committed:
refactor: separate therapist primary key from user relationship

- Add dedicated UUID primary key to therapists table
- Change user relationship to use user_id foreign key
- Update Makefile to use 'docker compose' v2 syntax
- Add Python 3.13 version specification

f1xgun committed:
feat: implement RBAC system with permissions and roles

- Add Permission and Role models with many-to-many relationships
- Create users_roles and roles_permissions association tables
- Implement RBAC repository layer with custom exceptions
- Add permission checking decorators (@require_permission, @require_any_permission)
- Refactor therapists repository into separate module with exceptions
- Add automatic role assignment: 'user' on registration, 'psychologist' on therapist creation
- Create PermissionCode and RoleCode enums for type-safe permission management
- Add RBAC endpoints for role assignment and removal
- Update Makefile to load database credentials from .env file
- Add migration for RBAC schema and data seeding
- Fix datetime timezone handling in appointments and users_roles tables
- Add TherapistCreateRequest schema and CRUD endpoints with permission checks
- Update README with quick start guide and environment setup

Alexanderbr1 commented on psychohelp/repositories/therapists/therapists.py:
```diff
@@ -0,0 +1,118 @@
+            
+            user_role_codes = {role.code for role in user.roles}
+            
+            if "psychologist" not in user_role_codes:
```
"psychologist" в const можно вынести

Alexanderbr1 commented on psychohelp/repositories/therapists/therapists.py:
```diff
@@ -0,0 +1,118 @@
+            if existing_therapist_result.scalar_one_or_none() is not None:
+                raise exc.TherapistAlreadyExistsException(user_id)
+            
+            therapist = Therapist(user_id=user_id, **therapist_data)
```
therapist надо переименовать (на паре обсуждали)

yakser commented on psychohelp/constants/__init__.py:
```diff
@@ -0,0 +1,11 @@
+"""
+Константы приложения
+"""
```
давай настолько очевидные комменты уберем)

yakser replied:
1 промпт и готово

yakser commented on psychohelp/constants/rbac.py:
```diff
@@ -0,0 +1,59 @@
+    
+    USER = "user"
+    PSYCHOLOGIST = "psychologist"
+    HEAD_OF_PSYCHOLOGISTS = "head_of_psychologists"
```
мб админ назовем? шо за хэд

Alexanderbr1 replied:
глава типо

Alexanderbr1 replied:
шо ты голова

yakser replied:
хпхапа да знаю

yakser commented on psychohelp/repositories/rbac/__init__.py:
```diff
@@ -0,0 +1,46 @@
+    "PermissionNotFoundException",
+    "RoleAlreadyAssignedException",
+    "RoleNotAssignedException",
+]
```
очень хочется это убрать, тут легко опечататься и как будто оно нафиг не надо

yakser commented on psychohelp/routes/controllers/roles.py:
```diff
@@ -1,57 +1,75 @@
 
 
-@router.get("/{user_id}", response_model=list[UserRole])
+@router.get("/{user_id}", response_model=list[RoleResponse])
```
эта ручка не нужна

yakser commented on psychohelp/services/rbac/permissions.py:
```diff
@@ -0,0 +1,203 @@
+)
+
+
+async def get_user_roles(user_id: UUID) -> list[Role]:
```
тоже будет не нужно если ручку уберем

yakser commented on psychohelp/repositories/rbac/rbac.py:
```diff
@@ -0,0 +1,159 @@
+    async with get_async_db() as session:
+        if isinstance(role_code, str):
+            try:
+                role_code = RoleCode[role_code] if role_code.isupper() else RoleCode(role_code)
```
вау!

yakser commented on psychohelp/repositories/rbac/rbac.py:
```diff
@@ -0,0 +1,159 @@
+
+async def get_role_by_code(role_code: str | RoleCode) -> Role | None:
+    async with get_async_db() as session:
+        if isinstance(role_code, str):
```
а есть вариант только RoleCode принимать?

yakser commented on psychohelp/repositories/rbac/rbac.py:
```diff
@@ -0,0 +1,159 @@
+        return list(result.scalars().all())
+
+
+async def assign_role_to_user(user_id: UUID, role_code: str) -> bool:
```
надо типизированный код роли принимать

yakser commented on psychohelp/repositories/rbac/rbac.py:
```diff
@@ -0,0 +1,159 @@
+            try:
+                role_enum = RoleCode[role_code] if role_code.isupper() else RoleCode(role_code)
+            except (ValueError, KeyError):
+                raise exc.RoleNotFoundException(role_code)
```
вау!

yakser commented on psychohelp/repositories/rbac/rbac.py:
```diff
@@ -0,0 +1,159 @@
+            if user is None:
+                raise exc.UserNotFoundException(user_id)
+            
+            role = next((r for r in user.roles if r.code.value == role_code), None)
```
а почему тут role_code не типизируем?)

yakser commented on psychohelp/repositories/rbac/rbac.py:
```diff
@@ -0,0 +1,159 @@
+            select(Permission).where(Permission.resource == resource)
+        )
+        return list(result.scalars().all())
+
```
не используется нигде в сервисах

yakser commented on psychohelp/schemas/roles.py:
```diff
@@ -0,0 +1,26 @@
+
+class RoleRemoveRequest(BaseModel):
+    """Схема запроса на удаление роли"""
+    role_code: str
```
тут можно типизировать role code

yakser commented on psychohelp/services/appointments/appointments.py:
```diff
@@ -0,0 +1,110 @@
+
+
+async def get_appointment_by_id(appointment_id: UUID):
+    return await repo_get_appointment_by_id(appointment_id)
```
а где тайп хинт для return value

yakser commented on alembic/versions/0434ee780edc_assign_user_role_to_existing_users.py:
давай не будем плодить кучу миграций,  тем более которые какие то фиксы делают

f1xgun committed:
fix: review issues

yakser commented on psychohelp/models/__init__.py:
```diff
@@ -1,17 +1,19 @@
+    "Permission",
+    "roles_permissions",
+    "users_roles",
 ]
```
можно тоже убрать

yakser commented on psychohelp/repositories/users.py:
```diff
@@ -52,7 +54,23 @@ async def create_user(
+            await session.flush()
+            
+            user_role_result = await session.execute(
+                select(Role).where(Role.code == "user")
```
можно из enum'а взять

f1xgun committed:
refactor: remove models/__init__.py and use direct imports

- Remove psychohelp/models/__init__.py as it only contained re-exports
- Update imports in rbac.py to use direct model imports
- Update imports in services/users/models.py to use direct User import
- Alembic and main.py already import modules directly, not classes
